### ISSUE TITLE
Repository type is not registered: path

### ISSUE THREAD

**Issue by johnbendi** [magento/magento2#2476]
When I add a repository of `type: path` to `composer.json` in root of installation, composer successfully installs the package from the `url` specified but `bin/magento` throws the above mentioned error during it's readiness check. 

Command line stacktrace:

```
We're sorry, an error occurred. Try clearing the cache and code generation directories. By default, they are: var/cache, var/di, var/generation, and var/page_cache.

  [Zend\ServiceManager\Exception\ServiceNotCreatedException]                                                           
  An abstract factory could not create an instance of magentosetupconsolecommandcronruncommand(alias: Magento\Setup\C  
  onsole\Command\CronRunCommand).                                                                                      

  [Zend\ServiceManager\Exception\ServiceNotCreatedException]                                                   
  An exception was raised while creating "Magento\Setup\Console\Command\CronRunCommand"; no instance returned  

  [Zend\Di\Exception\MissingPropertyException]                                                                        
  Missing instance/object for parameter readinessCheck for Magento\Setup\Console\Command\CronRunCommand::__construct  

  [Zend\ServiceManager\Exception\ServiceNotCreatedException]                                                           
  An abstract factory could not create an instance of magentosetupmodelcronreadinesscheck(alias: Magento\Setup\Model\  
  Cron\ReadinessCheck).                                                                                                

  [Zend\ServiceManager\Exception\ServiceNotCreatedException]                                              
  An exception was raised while creating "Magento\Setup\Model\Cron\ReadinessCheck"; no instance returned  

  [Zend\Di\Exception\MissingPropertyException]                                                                      
  Missing instance/object for parameter phpReadinessCheck for Magento\Setup\Model\Cron\ReadinessCheck::__construct  

  [Zend\ServiceManager\Exception\ServiceNotCreatedException]                                                           
  An abstract factory could not create an instance of magentosetupmodelphpreadinesscheck(alias: Magento\Setup\Model\P  
  hpReadinessCheck).                                                                                                   

  [Zend\ServiceManager\Exception\ServiceNotCreatedException]                                            
  An exception was raised while creating "Magento\Setup\Model\PhpReadinessCheck"; no instance returned  

  [Zend\Di\Exception\MissingPropertyException]                                                                      
  Missing instance/object for parameter composerInformation for Magento\Setup\Model\PhpReadinessCheck::__construct  

  [Zend\ServiceManager\Exception\ServiceNotCreatedException]                                                           
  An abstract factory could not create an instance of magentoframeworkcomposercomposerinformation(alias: Magento\Fram  
  ework\Composer\ComposerInformation).                                                                                 

  [Zend\ServiceManager\Exception\ServiceNotCreatedException]                                                     
  An exception was raised while creating "Magento\Framework\Composer\ComposerInformation"; no instance returned  

  [InvalidArgumentException]               
  Repository type is not registered: path 
```

**Comment by buskamuza** [comment]
Hi @johnbendi ,
could you, please, show an example of your `composer.json`?
As I can see from Composer's documentation (https://getcomposer.org/doc/04-schema.md#repositories) there is no type "path" in the list of repository types.

**Comment by dbritan** [comment]
@buskamuza please look here https://getcomposer.org/doc/05-repositories.md#path

**Comment by buskamuza** [comment]
@dbritan , got it, thanks.
"path" is added in Composer version 11, while Magento uses version 10. So Composer library throws this exception, when we try to work with newer version of `composer.json`.
For now, looks like `./bin/magento` is incompatible with format of `composer.json` newer then v.10 of Composer.
We'll review if it's possible to fix Magento CLI still depending on v.10 or should we update to v.11 of Composer.

**Comment by buskamuza** [comment]
Internal ticket MAGETWO-46666

**Comment by aboutvale** [comment]
I've the same problem described by @johnbendi.
This is a very interesting feature for extensions developers.

**Comment by mcspronko** [comment]
Hi Magento team, :)

Are there any updates for this issue? It will really help to start using "path" repository for module development.

Thanks.

**Comment by jacquesbh** [comment]
Actually you can. I just did it.

Here is my composer.json:

```js
{
    "name": "magento/project-community-edition",
    "description": "eCommerce Platform for Growth (Community Edition)",
    "type": "project",
    "version": "2.0.2",
    "license": [
        "OSL-3.0",
        "AFL-3.0"
    ],
    "repositories": [
        {
            "type": "composer",
            "url": "https://repo.magento.com/"
        },
        {
            "type": "path",
            "url": "../test",
            "options": {
                "symlink": true
            }
        }
    ],
    "require": {
        "jacques/test": "*",
        "magento/product-community-edition": "2.0.2",
        "composer/composer": "@alpha"
    },
    "require-dev": {
        "phpunit/phpunit": "4.1.0",
        "squizlabs/php_codesniffer": "1.5.3",
        "phpmd/phpmd": "@stable",
        "pdepend/pdepend": "2.2.2",
        "sjparkinson/static-review": "~4.1",
        "fabpot/php-cs-fixer": "~1.2",
        "lusitanian/oauth": "~0.3 <=0.7.0",
        "magento/module-bundle-sample-data": "100.0.*",
        "magento/module-theme-sample-data": "100.0.*",
        "magento/module-widget-sample-data": "100.0.*",
        "magento/module-catalog-sample-data": "100.0.*",
        "magento/module-customer-sample-data": "100.0.*",
        "magento/module-cms-sample-data": "100.0.*",
        "magento/module-catalog-rule-sample-data": "100.0.*",
        "magento/module-sales-rule-sample-data": "100.0.*",
        "magento/module-review-sample-data": "100.0.*",
        "magento/module-tax-sample-data": "100.0.*",
        "magento/module-sales-sample-data": "100.0.*",
        "magento/module-grouped-product-sample-data": "100.0.*",
        "magento/module-downloadable-sample-data": "100.0.*",
        "magento/module-msrp-sample-data": "100.0.*",
        "magento/module-configurable-sample-data": "100.0.*",
        "magento/module-product-links-sample-data": "100.0.*",
        "magento/module-wishlist-sample-data": "100.0.*",
        "magento/module-swatches-sample-data": "100.0.*",
        "magento/sample-data-media": "100.0.*",
        "magento/module-offline-shipping-sample-data": "100.0.*"
    },
    "config": {
        "use-include-path": true
    },
    "autoload": {
        "psr-4": {
            "Magento\\Framework\\": "lib/internal/Magento/Framework/",
            "Magento\\Setup\\": "setup/src/Magento/Setup/",
            "Magento\\": "app/code/Magento/"
        },
        "psr-0": {
            "": "app/code/"
        },
        "files": [
            "app/etc/NonComposerComponentRegistration.php"
        ]
    },
    "autoload-dev": {
        "psr-4": {
            "Magento\\Sniffs\\": "dev/tests/static/framework/Magento/Sniffs/",
            "Magento\\Tools\\": "dev/tools/Magento/Tools/",
            "Magento\\Tools\\Sanity\\": "dev/build/publication/sanity/Magento/Tools/Sanity/",
            "Magento\\TestFramework\\Inspection\\": "dev/tests/static/framework/Magento/TestFramework/Inspection/",
            "Magento\\TestFramework\\Utility\\": "dev/tests/static/framework/Magento/TestFramework/Utility/"
        }
    },
    "minimum-stability": "alpha",
    "prefer-stable": true,
    "extra": {
        "magento-force": "override"
    }
}
```

As you can see I have a "test" repository which use `path` as type.
On my machine I have the last version of composer.

So, maybe you can try a simple `composer self-update` and see if with the latest version of composer you're able to use a `path` repository?

**Comment by jacquesbh** [comment]
Oh my damn. I read not all the thing and I thought that it was just a composer usage… Let me check more. Sorry for the noise.

**Comment by jacquesbh** [comment]
_Note: I removed my useless two previous messages_

Ok, so. You can force composer to use a specific version for another… Let me explain with a really simple example.

Replace in your composer.json the require of `composer/composer` with this line:

```
"composer/composer": "dev-master as 1.0.0-alpha10"
```

It will say composer to use the `master` branch as version `1.0.0-alpha10` (version required by all Magento modules right now).

Hope this help :)

(Edit: Of course run a `composer update` after that…)

++

**Comment by mcspronko** [comment]
Hi @jacquesbh,

It now works as expected. Thank you.
I appreciate your fast response.

@benmarks thanks for tweet.

**Comment by buskamuza** [comment]
Magento depends on Composer version `1.0.0-beta1` now, so repository of type "path" should be supported.

**Comment by jacquesbh** [comment]
Can we close then?

**Comment by barryvdh** [comment]
This issue is still present in `v2.0.5`. The 2.0 branch still uses `1.0.0-alpha-10`: https://github.com/magento/magento2/blob/2.0/composer.json#L41

Also, there are now stable versions of both composer `1.0` and `1.1`, so perhaps upgrade to `1.1.x`?

**Comment by buskamuza** [comment]
Related internal ticket is MAGETWO-52770
@barryvdh , we're considering upgrading to the latest stable version. Thanks for the suggestion.

**Comment by kiborg** [comment]
I have the same issue. Magento CLI version 2.0.0 How to fix it?

**Comment by magento-engcom-team** [comment]
@johnbendi, thank you for your report.
The issue is already fixed in 2.2.0
